File: shared/geolocation.py
"""Canonical geolocation/site data shared by Orbit's Python services."""

from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import TypedDict


class GeolocationSite(TypedDict):
    id: str
    value: str
    name: str
    name_zh: str
    label_key: str


class GeolocationCatalog(TypedDict):
    version: int
    default_site: str
    sites: list[GeolocationSite]
    legacy_aliases: dict[str, str]


CATALOG_PATH = Path(__file__).with_name("geolocation.json")


@lru_cache(maxsize=1)
def get_geolocation_catalog() -> GeolocationCatalog:
    """Load the persisted catalog once and return a fresh immutable-by-convention value."""

    with CATALOG_PATH.open("r", encoding="utf-8") as source:
        return json.load(source)


GEOLOCATION_CATALOG = get_geolocation_catalog()
GEOLOCATION_SITES = tuple(site["value"] for site in GEOLOCATION_CATALOG["sites"])
DEFAULT_SITE = GEOLOCATION_CATALOG["default_site"]
LEGACY_SITE_ALIASES = GEOLOCATION_CATALOG["legacy_aliases"]


def normalize_site(value: object) -> str:
    """Return a canonical site value, falling back to the configured default."""

    site = str(value or "").strip()
    canonical = LEGACY_SITE_ALIASES.get(site, site)
    return canonical if canonical in GEOLOCATION_SITES else DEFAULT_SITE

File: frontend/src/geolocation.ts
import catalog from "../../shared/geolocation.json";

export interface GeolocationSite {
  id: string;
  value: string;
  name: string;
  name_zh: string;
  label_key: string;
}

export interface GeolocationCatalog {
  version: number;
  default_site: string;
  sites: readonly GeolocationSite[];
  legacy_aliases: Readonly<Record<string, string>>;
}

export const GEOLOCATION_CATALOG = catalog as GeolocationCatalog;
export const GEOLOCATION_SITES = GEOLOCATION_CATALOG.sites;
export const DEFAULT_SITE = GEOLOCATION_CATALOG.default_site;

export function normalizeSite(value: unknown): string {
  const site = String(value ?? "").trim();
  const canonical = GEOLOCATION_CATALOG.legacy_aliases[site] ?? site;
  return GEOLOCATION_SITES.some((option) => option.value === canonical)
    ? canonical
    : DEFAULT_SITE;
}

export function geolocationSite(value: unknown): GeolocationSite | undefined {
  const canonical = normalizeSite(value);
  return GEOLOCATION_SITES.find((option) => option.value === canonical);
}
